More fixes for the LLVM easyblock

Crivella opened the pull request:
## Fixes issues:

- #3708 
  - Fixed by cleaning up install directory by calling `self.make_installdir()` in the `prepare_step`
- #3709 
  - Fixed by attempting to rerurn the parallel build once if failed and once again in serial if failed again
  - Possible downside: more polluted logs for actually failed builds
- #3710
  - Implemented correct logic now also including runtimes libraries files
- #3711 
  - Added check based on version for expected folder structure 
- #3702 
  - Testing a dedicated `_prepare_runtimes_rpath_wrappers` method that can be ran also at the first build step
    - Tested with `bootstrap = False` `20.1.4` build (~minimal): https://github.com/easybuilders/easybuild-easyblocks/pull/3706#issuecomment-2844941653
    - Tested with `bootstrap = False` `20.1.4` build (all components): https://github.com/easybuilders/easybuild-easyblocks/pull/3706#issuecomment-2845350451
  - Needed to implement a different way to get the host triple that can be ran before the first build
    - Now taking it form the output of the CMake config logs instead of `glob`ing for the runtimes directory after the first build

## Other changes:

- Currently offloading is properly configured only with auto-detection of build targets. This PR allows the offloading to be configured independently of target detection
- Streamlined `GPU` and `ignore_patterns` for tests
- `SPIRV` moved to a non-experimental target (this might need to be enabled in a version specific way as this happened for versions `>=20.x`

Crivella commented:
Test report by @Crivella
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _LLVM-20.1.4.eb_ (partial log available at https://gist.github.com/Crivella/435fb2632aea9f7e52b253d605ad4b3c)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
crivella-desktop - Linux Ubuntu 22.04.5 LTS (Jammy Jellyfish), x86_64, 13th Gen Intel(R) Core(TM) i9-13900K, Python 3.11.12
See https://gist.github.com/Crivella/c0bbf29ec06904c2966dc9be834cbfa1 for a full test report.

Crivella commented:
@boegelbot please test @ jsc-zen3
EB_ARGS="LLVM-11.1.0-GCCcore-10.3.0.eb LLVM-16.0.6-GCCcore-12.3.0.eb LLVM-18.1.8-GCCcore-13.3.0-minimal.eb --installpath /tmp/$USER/ebpr-3706"

Crivella commented:
Test report by @Crivella
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _LLVM-20.1.3.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
crivella-desktop - Linux Ubuntu 22.04.5 LTS (Jammy Jellyfish), x86_64, 13th Gen Intel(R) Core(TM) i9-13900K, Python 3.11.12
See https://gist.github.com/Crivella/b4b01d722263f2c98849ad793508155b for a full test report.

Crivella commented:
Did one test locally with 20.1.4 and no bootstrap trying to minimize the build options as much as possible:

```
build_shared_libs = True

bootstrap = False
full_llvm = False
build_clang_extras = False
build_runtimes = True
build_lld = False
build_lldb = False
build_bolt = False
build_openmp = True
build_openmp_tools = False
build_openmp_offload = True
usepolly = False

python_bindings = True

build_targets = ['X86', 'AMDGPU']
```

but this lead to a new error in all offload tests:

```
# .---command stderr------------
# | /home/crivella/.local/easybuild/software/binutils/2.42-GCCcore-13.3.0/bin/ld: /home/crivella/.local/easybuild/build/LLVM/20.1.4/GCCcore-13.3.0-testruntimes/llvm.obj.1/bin/../lib/LLVMgold.so: error loading plugin: /home/crivella/.local/easybuild/build/LLVM/20.1.4/GCCcore-13.3.0-testruntimes/llvm.obj.1/bin/../lib/LLVMgold.so: cannot open shared object file: No such file or directory
# | clang: error: linker command failed with exit code 1 (use -v to see invocation)
# | /home/crivella/.local/easybuild/build/LLVM/20.1.4/GCCcore-13.3.0-testruntimes/llvm.obj.1/bin/clang-linker-wrapper: error: 'clang' failed
# | clang: error: linker command failed with exit code 1 (use -v to see invocation)
# `-----------------------------
# error: command failed with exit status: 1
```

for some reason linking with offloading and the GNU linker requires `LLVMgold.so` which is not being created.

I am wondering for this if we should:

- Open an issue and try to fix this later
- Just add a check and an error to always have `lld` on in non-minimal builds
- Both

The error being related to `offload` is probably a `>=19.x` problem but `LLVMgold.so` is in the cmake files also for previous versions, so this might need a more wide scope fix.

Thyre commented:
Given that `ld.gold` is being phased out starting with binutils 2.44, I would go the route of building `lld` if necessary, and using it instead of the gold linker. 

See #3703.

Crivella commented on easybuild/easyblocks/l/llvm.py:
```diff
@@ -858,31 +992,16 @@ def _para_test_step(self, parallel=1):
-        # TODO: Find a better way to either force the test to use the non wrapped compiler or to pass the flags
-        if build_option('rpath'):
-            setvar('CFLAGS', "%s %s" % (old_cflags, '-Wno-unused-command-line-argument'))
-            setvar('CXXFLAGS', "%s %s" % (old_cxxflags, '-Wno-unused-command-line-argument'))
```
This part is removed in favor of using the `LLVM-19.1.7_clang_rpathwrap_test.patch` in the easyconfigs

https://github.com/easybuilders/easybuild-easyconfigs/pull/22517/files#diff-2c19368d9e4d19b9a08b4803e398c65a3f599e810fc8cc265d548b747b028fa7

Crivella commented:
Test report by @Crivella
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _LLVM-20.1.4-GCCcore-13.3.0-testruntimes.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
crivella-desktop - Linux Ubuntu 22.04.5 LTS (Jammy Jellyfish), x86_64, 13th Gen Intel(R) Core(TM) i9-13900K, Python 3.11.12
See https://gist.github.com/Crivella/2d74d429acc38410d290e7eaa4b04bfd for a full test report.

Crivella commented:
Will try and rerun the previous test with everything but bootstrap as True

Crivella commented:
Test report by @Crivella
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _LLVM-20.1.4-GCCcore-13.3.0-nobootstrap.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
crivella-desktop - Linux Ubuntu 22.04.5 LTS (Jammy Jellyfish), x86_64, 13th Gen Intel(R) Core(TM) i9-13900K, Python 3.11.12
See https://gist.github.com/Crivella/12219cf6059c567a824879ad49181a45 for a full test report.

Thyre commented:
Test report by @Thyre
#### Overview of tested easyconfigs (in order)
 * **FAIL (unhandled exception: 'EB_LLVM' object has no attribute 'cuda_cc')Traceback (most recent call last):
  File "/home/jreuter/.local/lib/python3.10/site-packages/easybuild/main.py", line 137, in build_and_install_software
    (ec_res['success'], app_log, err_msg, err_code) = build_and_install_one(ec, init_env)
  File "/home/jreuter/.local/lib/python3.10/site-packages/easybuild/framework/easyblock.py", line 4609, in build_and_install_one
    result = app.run_all_steps(run_test_cases=run_test_cases)
  File "/home/jreuter/.local/lib/python3.10/site-packages/easybuild/framework/easyblock.py", line 4429, in run_all_steps
    self.run_step(step_name, step_methods)
  File "/home/jreuter/.local/lib/python3.10/site-packages/easybuild/framework/easyblock.py", line 4271, in run_step
    step_method(self)()
  File "/tmp/eb-ntmz3m7y/included-easyblocks-nt898huu/easybuild/easyblocks/llvm.py", line 677, in configure_step
    gpu_archs += ['sm_%s' % cc for cc in self.cuda_cc]
AttributeError: 'EB_LLVM' object has no attribute 'cuda_cc'
** _LLVM-19.1.1-GCCcore-13.3.0.eb_ 

Build succeeded for 0 out of 1 (1 easyconfigs in total)
zam226 - Linux Ubuntu 22.04, x86_64, 12th Gen Intel(R) Core(TM) i7-12700, Python 3.10.12
See https://gist.github.com/Thyre/e08cc64e07211b3715612e565a9f2898 for a full test report.

Crivella commented:
@Thyre https://github.com/easybuilders/easybuild-easyblocks/pull/3706/commits/64c219aa300c6a0dc316bbb8e4dd8100d18c6623 should fix https://github.com/easybuilders/easybuild-easyblocks/pull/3706#issuecomment-2857706388

Crivella commented:
Reverted fix for #3711 as the old project structure was still available (somehow i was using the `tags` instead of `releases` as the sources)
I've also added a way to use `start_dir` from the easyconfig to specify the starting directory which should make possible future changes (or usage with a non upstream llvm version using a different naming scheme) easier.

Thyre commented:
Test report by @Thyre
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _LLVM-19.1.1-GCCcore-13.3.0-amdgpu-nobootstrap.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
zam226 - Linux Ubuntu 22.04, x86_64, 12th Gen Intel(R) Core(TM) i7-12700, Python 3.10.12
See https://gist.github.com/Thyre/20c42f8d70de290085753b064ae03c5f for a full test report.

Thyre commented:
Test report by @Thyre
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _LLVM-19.1.1-GCCcore-13.3.0-amdgpu.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
zam226 - Linux Ubuntu 22.04, x86_64, 12th Gen Intel(R) Core(TM) i7-12700, Python 3.10.12
See https://gist.github.com/Thyre/92ee0ce0c0fdc30328d9c1ddaa8a5fdc for a full test report.

Thyre commented on easybuild/easyblocks/l/llvm.py:
```diff
@@ -933,24 +1010,23 @@ def test_step(self):
                 self._set_gcc_prefix()
                 self._create_compiler_config_file(self.cfg_compilers, self.gcc_prefix, self.final_dir)
+
+            self.ignore_patterns = ignore_patterns = self.cfg['test_suite_ignore_patterns'] or []
```
I would probably just use `self.ignore_patterns` here (and down below when adding patterns for NVPTX & AMDGPU). This would be a bit easier to read than this double assignment.

Crivella replied:
Done in https://github.com/easybuilders/easybuild-easyblocks/pull/3706/commits/65a02600bae5badb516ab53dbe3becf907f904c7

Thyre reviewed (commented):
Some minor comments. Looks good to me otherwise.

Thyre commented:
I would recommend that we do a test _without_ any CUDA compute capability or `amd_gfx_list` to check this as well. jsc-zen3 always checks with CUDA compute capabilities and my test ECs have `amd_gfx_list` set.
Looking at the code, this _shouldn't_ cause any issues, but I think it's a good idea to be a bit more safe...

Thyre commented:
Not directly related to this PR, but the EasyConfigs should make sure that we always use the `dlopen` variant for AMDGPU & NVPTX via a patch. LLVM does not require having CUDA or ROCm loaded to be built, and this gives us more flexibility (and prevents unexpected failures because a system ROCm / CUDA was pulled in somehow).

Or we set `LIBOMPTARGET_DLOPEN_PLUGINS` (named e.g. `LIBOMPTARGET_FORCE_DLOPEN_LIBHSA` for AMD in LLVM 18.1.8)

See e.g.:
https://github.com/llvm/llvm-project/blob/llvmorg-20.1.4/offload/plugins-nextgen/amdgpu/CMakeLists.txt#L11-L18

Crivella commented on easybuild/easyblocks/l/llvm.py:
```diff
@@ -813,10 +832,83 @@ def build_with_prev_stage(self, prev_dir, stage_dir):
             self.log.debug("Building %s", stage_dir)
             cmd = f"make {self.make_parallel_opts} VERBOSE=1"
-            run_shell_cmd(cmd)
+            res = run_shell_cmd(cmd, fail_on_error=False)
+            # Observed in 20.1.0, the build of the offloading tools can fail due to 'cstdint' file not found
+            # But will succeed if executed again with -j 1 (possible missing dependency in the CMake logic?)
+            # See https://github.com/llvm/llvm-project/issues/130783
+            if res.exit_code != EasyBuildExit.SUCCESS:
+                self.log.warning("Build failed, attempting again with parallel ON")
+                res = run_shell_cmd(cmd, fail_on_error=False)
+            if res.exit_code != EasyBuildExit.SUCCESS:
+                self.log.warning("Build failed, attempting again with parallel OFF")
+                cmd = "make -j 1 VERBOSE=1"
+                res = run_shell_cmd(cmd)
```
Agree but i think that would definitily be a standalone PRs requiring testing with every configuration of ECs (also would not be easy to check and reproduce the error as it is basically a race condition)

Crivella commented on easybuild/easyblocks/l/llvm.py:
```diff
@@ -813,10 +832,83 @@ def build_with_prev_stage(self, prev_dir, stage_dir):
+        #
+        # Here we create a mock binary for the current stage by copying the previous one, rpath-wrapping it and
+        # and than pass the rpath-wrapped binary to the runtimes build as the compiler.
+        #################################################
```
Done in https://github.com/easybuilders/easybuild-easyblocks/pull/3706/commits/35f9bf01f926ce16fef1c665ca32eeed4b7bd9ac

Crivella commented:
Test report by @Crivella
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _LLVM-20.1.4-GCCcore-13.3.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
crivella-desktop - Linux Ubuntu 22.04.5 LTS (Jammy Jellyfish), x86_64, 13th Gen Intel(R) Core(TM) i9-13900K, Python 3.11.12
See https://gist.github.com/Crivella/1f99ed88a6a00e2f2640b1ac2466b4ec for a full test report.

Crivella commented:
@boegel @bedroge I think this has been extensively tested, since you are already familiar with the new LLVM Easyblock would one of you be willing to give this a final check + merge?

Thyre commented:
Can you fix the f-string in line 1133? This isn't part of this PR, but broken anyway.

```
error_msg = "GCC installation path '{check_prefix}' does not match expected path '{gcc_prefix}'"
```

Crivella commented:
Rebased onto develop to fix merge conflict after

- #3681

Crivella commented:
@boegel @bedroge 

Tested again with one GCC version for every major release of LLVM for 18, 19, 20
- https://github.com/easybuilders/easybuild-easyconfigs/pull/21832#issuecomment-2883510458
- https://github.com/easybuilders/easybuild-easyconfigs/pull/22889#issuecomment-2881043761
- https://github.com/easybuilders/easybuild-easyconfigs/pull/22887#issuecomment-2880569005

The SYSTEM ones have the problem with the CI check, but I would not let it block this PR

I think now bot the ECs and the EBlock are good to go, and would try to merge them so the work on the toolchain can start being implemented

(As mentioned in https://github.com/easybuilders/easybuild-easyconfigs/pull/22887 there is also 20.1.5 released but it still does not have all the release files uploaded, in particular `llvm-20.1.5.src.tar.xz` is still not there yet)

ocaisa commented:
@Thyre If you have time, can you give a review on this, you are far more capable than me to do that (I'm happy to do a readthrough once you give the ok)

Thyre commented:
> @Thyre If you have time, can you give a review on this, you are far more capable than me to do that (I'm happy to do a readthrough once you give the ok)

Sure! I'll do another pass hopefully by the end of tomorrow. I'll give you a ping once I'm finished.

Thyre commented on easybuild/easyblocks/l/llvm.py:
```diff
@@ -408,28 +409,56 @@ def __init__(self, *args, **kwargs):
+        """Prepare step, modified to ensure install dir is deleted before building"""
+        super(EB_LLVM, self).prepare_step(*args, **kwargs)
+        # re-create installation dir (deletes old installation),
+        # Needed to unsure hardcoded rpath do not point to old installation during runtime builds and testing
```
```suggestion
        # Needed to ensure hardcoded rpath do not point to old installation during runtime builds and testing
```

Thyre commented on easybuild/easyblocks/l/llvm.py:
```diff
@@ -563,6 +592,17 @@ def configure_step(self):
         if self.cfg.parallel:
             self.make_parallel_opts = f"-j {self.cfg.parallel}"
 
+        # Moved here from the __init__ to ensure this easyblock can be used as a Bundle component
+        # https://github.com/easybuilders/easybuild-easyblocks/issues/3680
```
```suggestion
        # CMAKE_INSTALL_PREFIX and LLVM start directory are set here instead of in __init__ to 
        # ensure this easyblock can be used as a Bundle component, see
        # https://github.com/easybuilders/easybuild-easyblocks/issues/3680
```

Thyre commented on easybuild/easyblocks/l/llvm.py:
```diff
@@ -644,7 +684,27 @@ def configure_step(self):
-        super().configure_step(builddir=self.llvm_obj_dir_stage1, srcdir=src_dir)
+        output = super().configure_step(builddir=self.llvm_obj_dir_stage1, srcdir=src_dir)
+
+        # Get the LLVM HOST TRIPLE (e.g. x86_64-unknown-linux-gnu) from the output
```
```suggestion
        # Get LLVM_HOST_TRIPLE (e.g. x86_64-unknown-linux-gnu) from the output
```

Thyre reviewed (commented):
Minor comments regarding the comments in the EasyBlock. Otherwise LGTM.

Crivella commented on easybuild/easyblocks/l/llvm.py:
```diff
@@ -644,7 +684,27 @@ def configure_step(self):
+            if self.cfg['build_runtimes'] or self.cfg['bootstrap']:
+                raise EasyBuildError("`LLVM_HOST_TRIPLE` not found in the output of the configure step")
+            else:
+                self.log.warning("`LLVM_HOST_TRIPLE` not found in the output of the configure step")
```
Added https://github.com/easybuilders/easybuild-easyblocks/pull/3706/commits/37b554074da6e4f3f2b5a5ae53f7f199e47e8ea4

Thyre reviewed (approved):
LGTM

ocaisa commented on easybuild/easyblocks/l/llvm.py:
```diff
@@ -408,28 +409,56 @@ def __init__(self, *args, **kwargs):
+            if self.nvptx_target_cond:
+                if LooseVersion(self.version) < LooseVersion('20') and not cuda_cc_list:
+                    raise EasyBuildError(
+                        f"LLVM < 20 requires 'cuda-compute-capabilities' to build with {BUILD_TARGET_NVPTX}"
```
Try to be consistent with AMD below
```suggestion
                        f"LLVM < 20 requires 'cuda_compute_capabilities' to build with {BUILD_TARGET_NVPTX}"
```

ocaisa reviewed (changes requested):
Read through and just a minor comment

ocaisa reviewed (approved):
LGTM, thanks for the effort @Crivella and the detailed reviews @Thyre

ocaisa commented:
Builds have been tested in other PRs, and in particular https://github.com/easybuilders/easybuild-easyconfigs/pull/22903
